// @mft/moneyhub-api-client/dist/requests/affordability.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ({ config, request }) => {
    const { resourceServerUrl } = config;
    return {
        createAffordability: async ({ userId }, options) => request(`${resourceServerUrl}/affordability`, {
            cc: {
                scope: "affordability:read affordability:write",
                sub: userId,
            },
            method: "POST",
            options,
        }),
        getAffordability: async ({ userId, id }, options) => request(`${resourceServerUrl}/affordability/${id}`, {
            cc: {
                scope: "affordability:read",
                sub: userId,
            },
            method: "GET",
            options,
        }),
        getAllAffordability: async ({ userId, ...query }, options) => request(`${resourceServerUrl}/affordability`, {
            cc: {
                scope: "affordability:read",
                sub: userId,
            },
            method: "GET",
            searchParams: query,
            options,
        }),
    };
};
