// @mft/moneyhub-api-client/dist/requests/unauthenticated.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const query_string_1 = __importDefault(require("query-string"));
exports.default = ({ config, request }) => {
    const { resourceServerUrl, identityServiceUrl } = config;
    return {
        getGlobalCounterparties: (params = {}) => request(`${resourceServerUrl}/global-counterparties`, {
            searchParams: params,
        }),
        listConnections: (query) => request(`${identityServiceUrl}/oidc/.well-known/all-connections?${query && query_string_1.default.stringify(query)}`),
        listAPIConnections: (query) => request(`${identityServiceUrl}/oidc/.well-known/api-connections?${query && query_string_1.default.stringify(query)}`),
        listTestConnections: (query) => request(`${identityServiceUrl}/oidc/.well-known/test-connections?${query && query_string_1.default.stringify(query)}`),
        listPaymentsConnections: (query) => request(`${identityServiceUrl}/oidc/.well-known/payments-connections?${query && query_string_1.default.stringify(query)}`),
        listBetaConnections: (query) => request(`${identityServiceUrl}/oidc/.well-known/beta-connections?${query && query_string_1.default.stringify(query)}`),
        getOpenIdConfig: () => request(`${identityServiceUrl}/oidc/.well-known/openid-configuration`),
    };
};
